### ISSUE TITLE
[WICKET-7119] 5 unit tests are non-idempotent (passes in the first run but fails in subsequent runs in the same environment)

### ISSUE THREAD

**Issue by jira-importer** [jira-importer/jira-to-github-issues#51]
<p>List of non-idempotent tests:<br/>
org.apache.wicket.util.string.StringsTest#stripJSessionId<br/>
org.apache.wicket.resource.FileSystemResourceReferenceTest#testFileSystemResourceReferenceWithZip<br/>
org.apache.wicket.markup.repeater.AbstractPageableViewTest#cachedItemCount<br/>
org.apache.wicket.ComponentOnConfigureTest#order<br/>
org.apache.wicket.pageStore.AbstractPersistentPageStoreTest#rebindingAttributeDoesNotRemoveAllPages</p>

<p>Will open a PR to fix them</p>

---
<details><summary><i>Originally reported by <a title="Kaiyao Ke" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kaiyaok2">kaiyaok2</a>, imported from: <a href="https://issues.apache.org/jira/browse/WICKET-7119" target="_blank">5 unit tests are non-idempotent (passes in the first run but fails in subsequent runs in the same environment)</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Martin Tzvetanov Grigorov" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mgrigorov">mgrigorov</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2024-07-30T07:03:15+00:00
<li><b>imported</b>: 2025-11-27
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>kaiyaok2 opened a new pull request, #929:<br/>
URL: <a href="https://github.com/apache/wicket/pull/929" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/pull/929</a></p>

<p>   Fixes <a href="https://issues.apache.org/jira/browse/WICKET-7119" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/WICKET-7119</a></p>

<ol>
	<li>The Problem</li>
</ol>



<p>   Some unit tests are non-idempotent, as they pass in the first run but fail in the second run in the same environment. A fix is necessary since unit tests shall be self-contained, ensuring that the state of the system under test is consistent at the beginning of each test. In practice, fixing non-idempotent tests can help proactively avoid state pollution that results in test order dependency (which could cause problems under test selection, prioritization or parallelization).</p>


<ol>
	<li>Reproduce</li>
</ol>



<p>   Using the `NIOInspector` plugin that supports rerunning JUnit tests in the same environment. Use `org.apache.wicket.util.string.StringsTest#stripJSessionId` as an example:<br/>
   ```<br/>
   cd wicket-util<br/>
   mvn edu.illinois:NIOInspector:rerun -Dtest=org.apache.wicket.util.string.StringsTest#stripJSessionId</p>

<p>   ```</p>


<ol>
	<li>5 Non-Idempotent Tests & Proposed Fix
	<ol>
		<li>org.apache.wicket.util.string.StringsTest#stripJSessionId<br/>
   Reason: The `SESSION_ID_PARAM` field in `Strings`(see <a href="https://github.com/apache/wicket/blob/37c80e8c54533257c9fc5737424847a61e889b4e/wicket-util/src/main/java/org/apache/wicket/util/string/Strings.java#L74" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/blob/37c80e8c54533257c9fc5737424847a61e889b4e/wicket-util/src/main/java/org/apache/wicket/util/string/Strings.java#L74</a>) contains semicolon by default, but `stripJSessionId()` resets it to `"jsessionid"` (see <a href="https://github.com/apache/wicket/blob/37c80e8c54533257c9fc5737424847a61e889b4e/wicket-util/src/test/java/org/apache/wicket/util/string/StringsTest.java#L62" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/blob/37c80e8c54533257c9fc5737424847a61e889b4e/wicket-util/src/test/java/org/apache/wicket/util/string/StringsTest.java#L62</a>) when doing cleanup. As a result, `assertEquals(url, Strings.stripJSessionId(url + ";jsessionid=12345"));` will fail in the second test execution due to the missing semicolon.</li>
	</ol>
	</li>
</ol>


<p>   Error message of the test in the repeated run:<br/>
   ```<br/>
   org.opentest4j.AssertionFailedError: expected: <<a href="http://localhost/abc" class="external-link" target="_blank" rel="nofollow noopener">http://localhost/abc</a>> but was: <<a href="http://localhost/abc" class="external-link" target="_blank" rel="nofollow noopener">http://localhost/abc</a>;><br/>
   	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)<br/>
   	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)<br/>
   	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)<br/>
   	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)<br/>
   	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:177)<br/>
   	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1145)<br/>
   	at org.apache.wicket.util.string.StringsTest.stripJSessionId(StringsTest.java:43)<br/>
   ```<br/>
   Fix: Save the original value of the field `SESSION_ID_PARAM` before modifying it, and restore it to the saved value at the end of the test.</p>



<ol>
	<li>
	<ol>
		<li>org.apache.wicket.resource.FileSystemResourceReferenceTest#testFileSystemResourceReferenceWithZip<br/>
   Reason: `FileSystemAlreadyExistsException` occurs when calling `FileSystemResourceReference.getPath()` because the file system for the zip file is already mounted when the test is executed a second time. </li>
	</ol>
	</li>
</ol>


<p>   Error message in the repeated run:<br/>
   ```<br/>
   java.nio.file.FileSystemAlreadyExistsException: null<br/>
   	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:104)<br/>
   	at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:339)<br/>
   	at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:288)<br/>
   	at org.apache.wicket.resource.FileSystemJarPathService.getPath(FileSystemJarPathService.java:75)<br/>
   	at org.apache.wicket.resource.FileSystemResourceReference.getPath(FileSystemResourceReference.java:180)<br/>
   	at org.apache.wicket.resource.FileSystemResourceReference.getPath(FileSystemResourceReference.java:204)<br/>
   	at org.apache.wicket.resource.FileSystemResourceReferenceTest.testFileSystemResourceReferenceWithZip(FileSystemResourceReferenceTest.java:58)<br/>
   ```<br/>
   Fix: Check if the file system for the zip file is already available before creating a new one.</p>



<ol>
	<li>
	<ol>
		<li>org.apache.wicket.markup.repeater.AbstractPageableViewTest#cachedItemCount<br/>
   Reason: `View.getItemCount()` calls `View.internalGetItemCount()`, which is dependent on the static variable `count`. The test `cachedItemCount()` resets `count` to 8 without resetting it, so in the second test execution `view.getItemCount()` would be 8 instead of 5.</li>
	</ol>
	</li>
</ol>


<p>   Error message of one of the tests in the repeated run:<br/>
   ```<br/>
   org.opentest4j.AssertionFailedError: expected: <5> but was: <8><br/>
   	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)<br/>
   	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)<br/>
   	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)<br/>
   	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:166)<br/>
   	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:161)<br/>
   	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:632)<br/>
   	at org.apache.wicket.markup.repeater.AbstractPageableViewTest.cachedItemCount(AbstractPageableViewTest.java:44)<br/>
   ```<br/>
   Fix:  Add a `@BeforeEach` method to ensure `count` is reset to 5.</p>


<ol>
	<li>
	<ol>
		<li>org.apache.wicket.ComponentOnConfigureTest#order<br/>
   Reason: Similarly, the static atomic integers are not reset before / after the test, so they get accumulated across multiple test runs, causing assertion failures.</li>
	</ol>
	</li>
</ol>


<p>   Error message of one of the tests in the repeated run:<br/>
   ```<br/>
   org.opentest4j.AssertionFailedError: Page must be configured first! ==> expected: <0> but was: <6><br/>
   	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)<br/>
   	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)<br/>
   	at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)<br/>
   	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)<br/>
   	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:563)<br/>
   	at org.apache.wicket.ComponentOnConfigureTest.order(ComponentOnConfigureTest.java:54)<br/>
   ```<br/>
   Fix:  Add a `@BeforeEach` method to ensure the static atomic integers are reset to 0.</p>


<ol>
	<li>
	<ol>
		<li>org.apache.wicket.pageStore.AbstractPersistentPageStoreTest#rebindingAttributeDoesNotRemoveAllPages<br/>
   Reason: The constructor of `AbstractPersistentPageStore` checks if the input application name string already exists in the static concurrent map `STORES` (this would be the case in the second test execution), and throws `IllegalStateException` if so. </li>
	</ol>
	</li>
</ol>


<p>   Error message of one of the tests in the repeated run:<br/>
   ```<br/>
   java.lang.IllegalStateException: Store with key 'fooBar:' already exists.<br/>
   	at org.apache.wicket.pageStore.AbstractPersistentPageStore.<init>(AbstractPersistentPageStore.java:67)<br/>
   	at org.apache.wicket.pageStore.AbstractPersistentPageStoreTest$1.<init>(AbstractPersistentPageStoreTest.java:44)<br/>
   	at org.apache.wicket.pageStore.AbstractPersistentPageStoreTest.rebindingAttributeDoesNotRemoveAllPages(AbstractPersistentPageStoreTest.java:43)<br/>
   ```<br/>
   Fix:  Call `store.destroy()` to delete `'fooBar'` from the `STORES` map.</p>


<ol>
	<li>Verifying this change<br/>
   After the patch, running the tests repeatedly in the same environment will not lead to failures.</li>
</ol>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>martin-g merged PR #929:<br/>
URL: <a href="https://github.com/apache/wicket/pull/929" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/pull/929</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jira-bot">jira-bot</a>:</i>
<p>Commit dfa5713f9c60d76fb78d68c41df982236315488f in wicket's branch refs/heads/wicket-9.x from Kaiyao Ke<br/>
[ <a href="https://gitbox.apache.org/repos/asf?p=wicket.git;h=dfa5713f9c" class="external-link" target="_blank" rel="nofollow noopener">https://gitbox.apache.org/repos/asf?p=wicket.git;h=dfa5713f9c</a> ]</p>

<p><a href="https://issues.apache.org/jira/browse/WICKET-7119" title="5 unit tests are non-idempotent (passes in the first run but fails in subsequent runs in the same environment)" class="issue-link" data-issue-key="WICKET-7119"><del>WICKET-7119</del></a> fixed non-idempotent unit tests (#929)</p>

<p>(cherry picked from commit 1a55391e77e3bc6951a59ac99c6bf23c8b66d40b)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>martin-g commented on PR #929:<br/>
URL: <a href="https://github.com/apache/wicket/pull/929#issuecomment-2257619329" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/wicket/pull/929#issuecomment-2257619329</a></p>

<p>   Thank you, @kaiyaok2 !</p>
